onnxruntime/core/providers/stvm/stvm_execution_provider.cc

```diff
@@ -0,0 +1,535 @@
+#include "xpu_data_transfer.h"
+#include "stvm_allocator.h"
+#include "stvm_utils.h"
+#include "stvm_api.h"
+
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+
+// Information to construct kernel function state.
+struct STVMFuncState {
+  AllocateFunc allocate_func = nullptr;
+  DestroyFunc release_func = nullptr;
+  AllocatorHandle allocator = nullptr;
+  tvm::runtime::Module* module = nullptr;
+  std::function<tvm::runtime::Module*(std::string func_name, const std::vector<std::vector<int64_t>>& input_shapes)> compiler = nullptr;
+};
+
+class STVMRunner {
+  public:
+    using TVMTensorShape = std::vector<int64_t>;
+    using TVMTensorShapes = std::vector<TVMTensorShape>;
+    using InputsInfoMap = std::map<size_t, TVMTensorShape>;
+    using ORTGraphNodes = std::vector<const NodeArg*>;
+
+    STVMRunner() = delete;
+    ~STVMRunner() = default;
+
+    STVMRunner(StvmExecutionProvider* ep,
+               const std::string& name,
+               const Graph& graph) {
+        // Extract input shapes
+        const ORTGraphNodes& all_nodes = graph.GetInputsIncludingInitializers();
+        TVMTensorShapes input_shapes;
+        size_t indx = 0;
+        if (ep->info_.freeze_weights) {
+          for (const auto* node : all_nodes) {
+            const auto& node_name = node->Name();
+            if(!graph.IsInitializedTensor(node_name)) {
+              TVMTensorShape ishape;
+              if(!ep->info_.input_shapes.empty() &&
+                  ep->info_.input_shapes.count(node_name)) {
+                ishape = ep->info_.input_shapes[node_name];
+                inputs_info_[indx] = ishape;
+                update_output_shapes_ = true;
+              } else {
+                getTensorInfo(*node->Shape(), ishape, indx);
+              }
+              input_shapes.emplace_back(ishape);
+            }
+            ++indx;
+          }
+        } else {
+          for (const auto* node : all_nodes) {
+            const auto& node_name = node->Name();
+            TVMTensorShape ishape;
+            if(!ep->info_.input_shapes.empty() &&
+                ep->info_.input_shapes.count(node_name)) {
+              ishape = ep->info_.input_shapes[node_name];
+              inputs_info_[indx++] = ishape;
+              update_output_shapes_ = true;
+            } else {
+              getTensorInfo(*node->Shape(), ishape, indx++);
+            }
+            if(!graph.IsInitializedTensor(node_name)) {
+              input_shapes.emplace_back(ishape);
+            }
+          }
+        }
+
+        // Get module from tvm
+        mod_ = ep->CompileFunc(name, input_shapes);
+
+        // Prepare draft for output tvm tensors
+        const ORTGraphNodes& ort_outputs_info = graph.GetOutputs();
+        size_t num_outputs = ort_outputs_info.size();
+
+        if (update_output_shapes_) {
+          stvm::TVMGetOutputShapes(*mod_, num_outputs, output_shapes_);
+        } else {
+          for (auto i = 0u; i < num_outputs; i++) {
```

Comment:
0u unsigned, num_outputs is signed?

Reply:
See line 91, num_outputs is size_t (unsigned long)